doc/tomtom.1.md
% tomtom(1) Tomtom %TOMTOM_VERSION%
% Gabriel Filion <lelutin@gmail.com>
% 2010-03-25

# NAME

tomtom - CLI interface to Tomboy or Gnote via DBus

# SYNOPSIS

tomtom (h|--help) [command]

tomtom \<command\> [options...]

# DESCRIPTION

`tomtom` is an interface to note-taking applications `Tomboy notes` and `Gnote`
that uses DBus to communicate with them. It presents a command-line interface
and tries to be as simple to use as possible. Different actions can be taken to
interact with Tomboy or Gnote. Actions are simple to create, making the
application easily extensible.

To obtain details about how to invoke subcommands, use the "-h" option followed
by the subcommand name.

# SUBCOMMANDS

`list`
:   List note names and other useful information about them.
`display`
:   Show the contents of one or more notes.
`search`
:   Perform a textual search on your notes' contents.
`delete`
:   Delete one or more notes, or an entire book.
`version`
:   Show version tags from Tomtom and the application it connects to.

# EXAMPLES

List the 10 notes last modified
:   tomtom list -n 10
Display the contents of note "thing X"
:   tomtom display "thing X"
Search for "needle" in notes of the "haystack" book
:   tomtom search -b haystack needle
Delete notes with the tag "old_and_busted"
:   tomtom delete -t old_and_busted

# SEE ALSO

See the *README.markdown* file for more information.

The project's page is situated at <http://github.com/lelutin/tomtom>.

# BUGS

To report issues, go to the "issues" section of the projet's page:
<http://github.com/lelutin/tomtom/issues>.


CHANGELOG
Changes since 0.2:

* New "version" action shows information about the note application's version
* Fixed encoding issue on output
* New option -v|--version (without an action) displays information about Tomtom
* Actions are now setuptools plugins: makes it possible to compress the .egg
* Libraries of options make it easier to centralize options for actions
* Moved action code away from core.Tomtom
* Reorganized code in core.py -- there is one less class
* Created a "help" pseudo-action as an alias to 'action -h'
* Added configuration files for easily running nosetests and pylint
* Tighter control over what is done during tests
* Added Gnote support over DBus as well
* Tomtom now searches for global and user configuration files
* Improved test coverage. Unit tests now cover 100% of the code
* Refined and standardized actions' usage text
* Cleaned up the test directory
* Added a "delete" action to delete notes
* Lots of code refactoring to make it easier to work with

Changes since 0.1:

* Fixed NoteNotFound handling (issue #1)
* Fixed searching for non-ascii characters (issue #2)
* Added KeyboardInterrupt (Ctrl-C) handling (issue #8)
* Improved coverage by added tests that were missing
* Refactored tests for better organization and better function call
  verification
* Added automating installation with setuptools
* Removed -a option from list action. Default behavior is now to list all
  notes. Number of lists can be restricted with -n option.
* Added selection by tag (issue #4)
* Templates are now filtered by default (issue #3)
* No more test runner code. Now using nosetests



tomtom/tomtom.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
###############################################################################
#
# Copyright (c) 2009, Gabriel Filion
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
#     * Redistributions of source code must retain the above copyright notice,
#       this list of conditions and the following disclaimer.
#     * Redistributions in binary form must reproduce the above copyright
#       notice,
#     * this list of conditions and the following disclaimer in the
#       documentation and/or other materials provided with the distribution.
#     * Neither the name of the copyright holder nor the names of its
#       contributors may be used to endorse or promote products derived from
#       this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
#
###############################################################################
#
# Inspired by : http://arstechnica.com/open-source/news/2007/09/using-the-tomboy-d-bus-interface.ars
#
"""Usage: tomtom.py (-h|--help) [action]
       tomtom.py <action> [-h|--help] [options]

Tomtom is a command line interface to the Tomboy note taking application.

Options depend on what action you are taking. To obtain details on options
for a particular action, combine -h or --help and the action name.

Here is a list of all the available actions:
  list    : listing all or the 10 latest notes
  display : displaying one or more notes at a time
  search  : searching for text in all notes or a specific list of notes

"""
import sys
import os
import optparse

from utils import *

def action_list_notes(args):
    """Use the tomtom object to list notes.

    This action prints modification date, title and tags of notes to the
    screen.

    Arguments:
        args -- A list composed of action and file names

    """
    parser = optparse.OptionParser(usage="%prog list [-h|-a]")
    parser.add_option("-a", "--all",
        dest="full_list", default=False, action="store_true",
        help="List all the notes")

    (options, file_names) = parser.parse_args(args)

    tomboy_interface = Tomtom()

    if options.full_list:
        print tomboy_interface.list_notes()
    else:
        print tomboy_interface.list_notes(count_limit=10)

def action_print_notes(args):
    """Use the tomtom object to print the content of one or more notes.

    This action fetches note contents and displays them to the screen.

    Arguments:
        args -- A list composed of action and file names

    """
    parser = optparse.OptionParser(usage="%prog display [-h] [note_name ...]")
    #parser.add_option("-a", "--all",
    #    dest="full_list", default=False, action="store_true",
    #    help="Display all the notes")

    (options, file_names) = parser.parse_args(args)

    if len(file_names) <= 0:
        print >> sys.stderr, \
            "Error: You need to specify a note name to display it"
        return

    tomboy_interface = Tomtom()

    try:
        print tomboy_interface.get_display_for_notes(file_names)
    except NoteNotFound, e:
        print >> sys.stderr, """Note named "%s" not found.""" % e

def action_search_in_notes(args):
    """Use the tomtom object to search for some text within notes.

    This action performs a textual search within notes and reports the results
    to the screen.

    Arguments:
        args -- list of arguments decomposed into action and file names

    """
    parser = optparse.OptionParser(
        usage="%prog search [-h] <search_pattern> [note_name ...]"
    )

    (options, file_names) = parser.parse_args(args)

    if len(file_names) < 1:
        print >> sys.stderr, \
            "Error: You must specify a pattern to perform a search"
        return

    tomboy_interface = Tomtom()

    search_pattern = file_names[0]
    note_names = file_names[1:]

    results = tomboy_interface.search_for_text(
        search_pattern=search_pattern,
        note_names=note_names
    )
    for result in results:
        print "%s : %s : %s" % \
            ( result["title"], result["line"], result["text"] )

# This dictionary is used to dispatch the actions and to list them for the -h
# option
#TODO find a better way to dispatch to really make implementing a new action
#simple as pie
available_actions = {
    "list": action_list_notes,
    "display": action_print_notes,
    "search": action_search_in_notes,
}

def main():
    """Application entry point.

    Checks the first parameters for general help argument and dispatches the
    actions.

    """
    if len(sys.argv) < 2:
        # Use the docstring's first [significant] lines to display usage
        usage_output =  (os.linesep * 2).join([
            os.linesep.join( __doc__.splitlines()[:3] ),
            "For more details, use option -h"])
        print usage_output
        return

    action = sys.argv[1]
    arguments = sys.argv[2:]

    if action in ["-h", "--help"]:
        if sys.argv[2:]:
            # Switch -h and action name and continue as normal
            arguments = [sys.argv[2], action] + sys.argv[3:]
            action = sys.argv[2]
        else:
            # Use the script's docstring the for basic help message
            print __doc__[:-1]
            return

    if action not in available_actions:
        print >> sys.stderr, \
            """%s: %s is not a valid action. """ % (sys.argv[0], action) + \
            """Use option -h for a list of available actions."""
        return

    perform_with = available_actions[action]
    perform_with(arguments)

if __name__ == "__main__":
    main()

